### PR TITLE
Remove conversion from pointer to reference and back again for ReverseComparator.

### PR DESCRIPTION
**Pull request by nickshanks**
# Refactor

## Summary
I tried to simplify this function down to a simple line, but failed.  
This is the only thing left from that effort that I think is worth keeping.

## Testing Done
Sorted and reverse-sorted all table columns.

### CODE REVIEW

**Comment by TheGiraffe3** [comment]
Can you give a bit more information in the title? For non-C++ people it's a bit hard to figure out what you're doing.

File: source/PlayerInfoPanel.cpp [hunk 1]
```diff
@@ -138,19 +138,19 @@ namespace {
 	}
 
 	// Reverses the argument order of the given comparator function.
-	InfoPanelState::ShipComparator &GetReverseCompareFrom(InfoPanelState::ShipComparator &f)
+	InfoPanelState::ShipComparator &GetReverseCompareFrom(InfoPanelState::ShipComparator f)
```
Comment (petervdmeer) [inline source/PlayerInfoPanel.cpp:141]: This was a pointer to ShipComparator, so I expected the parameter to be `*f`, but both `f` as well as `*f` seems to work. Does C++ perform some automated conversion magic here? (And we are not accidentally calling by value?)

Comment (TheGiraffe3) [reply source/PlayerInfoPanel.cpp:141]: Isn't that what he is doing, removing the comparators?

Comment (RisingLeaf) [reply source/PlayerInfoPanel.cpp:141]: Comparatir is a shared_ptr

Comment (nickshanks) [reply source/PlayerInfoPanel.cpp:141]: > This was a pointer to ShipComparator, so I expected the parameter to be `*f`, but both `f` as well as `*f` seems to work (I tested both). Does C++ perform some automated conversion magic here? (And we are not accidentally calling by value?)

I *think* I am removing a call-by-value and just passing a function pointer instead. But it's a bit hard to follow (then again, it is C++, so…)

I am pretty sure the code as suggested in this PR is just passing around function pointers and not doing anything trippy.

**Comment by warp-core** [comment]
`shipComparator` is of type `InfoPanelState::ShipComparator *`.
`InfoPanelState::ShipComparator` is itself an alias of `bool(const std::shared_ptr<Ship> &, const std::shared_ptr<Ship> &)`. A function.
But, you can't pass a function by value. Passing by value involves copy-constructing a new object of the same type with the original. How you gonna copy-construct a function?
It just passes the address of the function.

We can see this in the disassembly view while ebugging:
image
This copies 8 bytes (a 64-bit address) at the address of `shipComparator` (so the contents of that variable, which is the address of the function it points to) into register `rcx` (this'll be where the next function expects to get its parameter from). It then calls the function, and then copies 8 bytes from register `rax` (this'll be where the other function puts its return value) into the address of `shipComparator`, updating the address it points to.

image
Here, we can see what the called function does.
It starts by subtracting 16 from the value in register `rsp` (the stack pointer) and storing the result in that same register.
This is creating the stack frame for this function. It will have 16 bytes available to it.
It then copies the value in register `rcx` (where the address pointed to by `shipComparator` was placed before this function was called) into 8 bytes at the address pointed to by `rsp`. Putting the parameter on the stack.
Note that this is clearly pass by value; any changes made to this value will not immediately be present in the variable owned by the caller.
It then checks to see if the passed value is equal to the address of the `CompareName` function. It starts by loading the address of that function into register `rax`. It then compares the value in register `rax` with the 8 bytes at the address pointed to by the value in register `rsp`. If they are not equal, it jumps to the next comparison in this `if` block. If they are equal, it loads the address of the reverse compare function into register `rax`, copies that address from register `rax` to the 8 bytes that are 8 bytes up from the address pointed to by the value in register `rsp` (the second 8 bytes in the stack frame for this function) and then jumps elsewhere in this function.

Here's where it goes, the end of the function:
image
It copies the value from the stack frame into register `rax`, increases the value in register `rsp` by 16 (back to where the stack frame for hte calling function is) and then returns.

Here's the disassembly when I remove the dereference operator when making hte call and change the function to accept a pointer to an `InfoPanelState::ShipComparator`, instead of a reference to it.
image
image
image
Looks fairly identical to me.

And, with this PR:
image
image
image
Again, fairly identical.

So, basically, just choose whichever version is most understandable, because they aren't any different under the hood.

**Comment by warp-core** [comment]
Here's the assembly when comipled with gcc:
Vanilla:
image
image
image

This PR:
image
image
image

I don't think this is as easy to follow as the clang version, but it still seems like the same stuff is happening in both versions.

**Comment by petervdmeer** [comment]
> I think the best approach would be to not dereference the pointer when calling the function, for the function to accept a pointer to `InfoPanelState::ShipComparator` and for the comparisons to use the address-of operator on the functions that are invovled.
> I think this way makes it clearest that the code is working with function addresses.

Yes, that seems most clear.

**Comment by nickshanks** [comment]
> and for the comparisons to use the address-of operator on the functions that are invovled

I disagree with this. A bare function name in the code is a pointer to that function.
`&someFunc` reads to me as 'address of pointer to (the entry point of) `someFunc()`'.

When passed a function pointer, C/C++ code can call it by using `(*someFunc)(arg1, arg2)`. This makes it clear that the intrinsic/underlying data type of someFunc IS a pointer, and not the function itself.

I want the code to read as it executes: to compare function pointers with one another.
Do you read it differently?

**Comment by petervdmeer** [comment]
This code is about refactoring; it doesn't add functions or fixes bugs, it only changes existing code to do still exactly the same as before. The bar for merging refactoring code is quite high, because the new code needs to be significantly more readable and clear than the old code.

The discussion here tells me that this PR is not achieving this significant improvement, so I'm going to close this PR.
